### ISSUE TITLE
REACT_APP_* environment variables are lost when using ionic serve and ionic build

### ISSUE THREAD

**Issue by gdrbyKo1** [ionic-team/ionic-cli#4333]
**Description:**
react-scripts supports embedding the values of certain environment variables at build time: https://create-react-app.dev/docs/adding-custom-environment-variables/
however, when a react project is started with `ionic serve` or built with `ionic build`, the environment variables are ignored.

**Steps to Reproduce:**

1. Make a `console.log(process.env.REACT_APP_MYVAR)` call somewhere in your ionic react app
2. Execute `REACT_APP_MYVAR=foo ionic serve` or `REACT_APP_MYVAR=foo ionic build`
3. Run the app and observe how the console.log call outputs "undefined" instead of "foo"
4. Execute `REACT_APP_MYVAR=foo npm start` or `REACT_APP_MYVAR=foo npm run build`
5. Run the app and observe how the console.log call now properly outputs "foo" rather than "undefined"

**My `ionic info`:**

```
Ionic:

   Ionic CLI       : 5.4.16 (C:\Users\foo\AppData\Roaming\npm\node_modules\ionic)
   Ionic Framework : @ionic/react 4.11.10

Capacitor:

   Capacitor CLI   : 1.3.0
   @capacitor/core : 1.3.0

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v13.9.0 (C:\Program Files\nodejs\node.exe)
   npm    : 6.13.7
   OS     : Windows 10
```

**Comment by bnowlin** [comment]
I have the same issue. Even when using .env files.

Ionic:
   Ionic CLI       : 5.4.6 (C:\Users\me\AppData\Roaming\npm\node_modules\ionic)
   Ionic Framework : @ionic/react 5.0.0
Capacitor:
   Capacitor CLI   : 1.3.0
   @capacitor/core : 1.3.0
Utility:
   cordova-res : not installed
   native-run  : not installed
System:
   NodeJS : v12.13.1 (C:\Program Files\nodejs\node.exe)
   npm    : 6.13.4
   OS     : Windows 10

**Comment by mhartington** [comment]
👋 Hey folks! Can you try updating your CLI the latest release? 

```
npm uninstall -g ionic
npm install -g @ionic/cli 
```

I've tested this with 6.1 and all is working well for me.

**Comment by dwieeb** [comment]
Ionic CLI 6.2.1 should work! Thanks for the report!

**Comment by dahalsantosh007** [comment]
it doesn't work for me. I tried both ionic serve and npm run start. Both don't work for me

**Comment by dahalsantosh007** [comment]
I tried getting 6.2.1 I wasn't able to do it. I am not sure how to completely solve this problem. is there anyone who can help me

**Comment by MatHuf** [comment]
Experiencing this issue with 6.7.0
Same steps to reproduce as the original reporter.

**Comment by dwieeb** [comment]
@MatHuf I'm not able to reproduce. Can you post your `ionic info`?

**Comment by MatHuf** [comment]
@dwieeb Please disregard. I was trying to set the NODE_ENV variable and only just learned that CRA prevents this from being changed. Sorry to waste your time.

**Comment by dwieeb** [comment]
No worries! Glad it's working.

**Comment by solracid** [comment]
I am experiencing the same issue, I defined in:
`.env.local: 
    REACT_APP_API_BACKEND_ENDPOINT=http://localhost:1339/`
`.env.prod:
    REACT_APP_API_BACKEND_ENDPOINT=https://mydevapi-backend.herokuapp.com/`

I added `console.log(`env-settings:: ${JSON.stringify(process.env)}`)`. On running my app on my phone with Android Studio:
`2020-06-29 13:27:07.716 27044-27044/io.ionic.starter I/Capacitor/Console: File: http://localhost/static/js/main.0cb3591f.chunk.js - Line 1 - Msg: env-settings: {"NODE_ENV":"production","PUBLIC_URL":"","REACT_APP_API_BACKEND_ENDPOINT":"http://localhost:1339/"}`

My Ionic info:

```
Ionic:

   Ionic CLI       : 6.10.1 (C:\Users\bar\AppData\Roaming\npm\node_modules\@ionic\cli)
   Ionic Framework : @ionic/react 5.0.7

Capacitor:

   Capacitor CLI   : 2.0.0
   @capacitor/core : 2.1.2

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v12.16.0 (C:\Program Files\nodejs\node.exe)
   npm    : 6.14.5
   OS     : Windows 10
```

**Comment by imhoffd** [comment]
According to [these docs](https://create-react-app.dev/docs/adding-custom-environment-variables/#what-other-env-files-can-be-used) you're supposed to use `.env.production`, not `.env.prod`.

**Comment by jemantilla** [comment]
Hi Im also experiencing this 

```Ionic:

   Ionic CLI       : 6.10.1 (/Users/neil/.nvm/versions/node/v10.16.3/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 5.0.7

Capacitor:

   Capacitor CLI   : 2.2.1
   @capacitor/core : 2.3.0

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v10.16.3 (/Users/neil/.nvm/versions/node/v10.16.3/bin/node)
   npm    : 6.13.6
   OS     : macOS Mojave```

**Comment by imhoffd** [comment]
@jemantilla Can you post reproduction steps?

**Comment by looklookson** [comment]
I use ionic-cli 6.11.0, but the problem still there.

**Comment by imhoffd** [comment]
Again, without reproduction steps I'll be unable to help. Example repos demonstrating the problem are the best way to report bugs. Without a valid reproduction, I'm going to lock this issue.

**Comment by jemantilla** [comment]
To reproduce
Steps to reproduce the behavior:

1. Build mobile using ```cross-env REACT_APP_APPENV=development react-scripts build && ionic capacitor copy ios && ionic capacitor open ios```
1. Try to print env variable REACT_APP_APPENV using ```console.log(process.env)```
Result: ```{"NODE_ENV":"production","PUBLIC_URL":""}```

**Comment by imhoffd** [comment]
@jemantilla `ionic capacitor copy ios` also runs a build, so the `react-scripts build` step is unnecessary. If you use `ionic capacitor build`, it does all those steps (build, copy, and open). Also, your environment variable is only being set for the first command. That's how bash works. You can test this with `REACT_APP_APPENV=development true && env` (the env variable is set for the `true` command, but not for `env` command).

TLDR: don't split up the build steps, use higher-level commands:

```
REACT_APP_APPENV=development ionic capacitor build 
```

OR just use [`.env` files](https://create-react-app.dev/docs/adding-custom-environment-variables/#adding-development-environment-variables-in-env), they're better.

**Comment by jemantilla** [comment]
Woah, this ^ fixed my problem. Thanks!

**Comment by marksyzm** [comment]
I found that also the PUBLIC_URL is always wiped, and often the `REACT_APP_*` name space is all I can use here too. Hopefully that's of some use.

**Comment by FernetB** [comment]
I'm experiencing the same issue

Script:
`
"start": "set REACT_APP_API_ENV=local&&ionic serve" `

`"start": "set 'REACT_APP_API_ENV=local'&&ionic serve" `

Both of this does not work

But this works `"start": "set REACT_APP_API_ENV=local&&react-scripts start"`
```
Ionic:

   Ionic CLI       : 6.18.1 (C:\Users\PC\AppData\Roaming\nvm\v14.17.0\node_modules\@ionic\cli)
   Ionic Framework : @ionic/react 5.8.5

Capacitor:

   Capacitor CLI      : 3.0.2
   @capacitor/android : 3.2.4
   @capacitor/core    : 3.0.2
   @capacitor/ios     : not installed

Utility:

   cordova-res : 0.15.1
   native-run  : 1.5.0

System:

   NodeJS : v14.17.0 (C:\Program Files\nodejs\node.exe)
   npm    : 6.14.13
   OS     : Windows 10
```

**Comment by darron1217** [comment]
I'm using `env-cmd` but only `REACT_APP_*` variables are being lowercased.

**Comment by FernetB** [comment]
Any updates? This is still an issue in `Windows 10`, on linux works perfectly.
@imhoffd 

I'm doing: `set 'REACT_APP_IS_NATIVE=true'&&set 'REACT_APP_API_ENV=stage'&&ionic cap sync android` if it helps

also this works:
`"set 'REACT_APP_API_ENV=local'&&set PORT=8100&&react-scripts start"`
but this doesn't:
`"set 'REACT_APP_API_ENV=local'&&ionic serve"`

```
Ionic:

   Ionic CLI       : 6.19.0
   Ionic Framework : @ionic/react 5.9.1

Capacitor:

   Capacitor CLI      : 3.0.2
   @capacitor/android : 3.2.4
   @capacitor/core    : 3.0.2
   @capacitor/ios     : not installed

Utility:

   cordova-res : 0.15.1
   native-run  : 1.5.0

System:

   NodeJS : v14.17.0
   npm    : 6.14.13
   OS     : Windows 10
```

**Comment by piotr-cz** [comment]
On windows, environment variables keys are *normalized* be in lowercase here: https://github.com/ionic-team/ionic-cli/blob/2620059648f6692540b17e4e2279d2b17507d1ae/packages/%40ionic/utils-process/src/index.ts#L25-L35

Maybe it was an attempt to prevent duplicates, however the fact is that it's not possible to access environment variables in final builds unless one define them in lowercase format.

In my opinion it would make more sense to normalize keys to uppercase format, at least it's a convention to use environment variables this way.

**Comment by darron1217** [comment]
Here's my workaround.
I build web content first, then run native build (with `--no-build` option).

For example, this is the script I use for Android Development environment.
`env-cmd -f .env.development react-scripts build && ionic cap build android --no-build`

I hope this can help someone...

**Comment by piotr-cz** [comment]
My workaround is to reference shell environment variables in .env file (as described in CRA docs > [Adding Custom Environment Variables](https://create-react-app.dev/docs/adding-custom-environment-variables/#expanding-environment-variables-in-env)):

```yml
# .gitlab-ci.yml
build:
  script:
    - export VERSION=$(scripts/get-semver.sh $CI_COMMIT_TAG)
    - npm run build
```

```ini
# .env
REACT_APP_VERSION=${VERSION}
```

**Comment by rkrisztian** [comment]
I have the same issue as what @piotr-cz found, particularly using ionic 5 (have not tried ionic 6 yet):
https://github.com/ionic-team/ionic-cli/issues/4333#issuecomment-1108499369

We use git-bash on Windows (we are actually multi-platform with Linux and Mac), and environment variables are important for us, e.g., we use them in `.npmrc` to provide authentication tokens. So **I'd rather not normalize anything** on Windows. The Android Gradle build started by ionic also needs to use some inherited env vars.
